vortex-array/src/compute/filter.rs

```diff
@@ -103,209 +105,426 @@ fn filter_impl(array: &ArrayData, mask: FilterMask) -> VortexResult<ArrayData> {
+                slices.push((start, prev + 1));
 
-    fn size_hint(&self) -> (usize, Option<usize>) {
-        let remaining = self.trusted_len - self.index;
-        (remaining, Some(remaining))
-    }
-}
+                return slices;
+            }
 
-/// Safety: BitIndexIterator is TrustedLen because it knows its total length.
-unsafe impl TrustedLen for BitIndexIterator<'_> {}
-impl ExactSizeIterator for BitIndexIterator<'_> {}
+            vortex_panic!("No mask representation found")
+        })
+    }
 
-pub enum FilterIter<'a> {
-    // Slice of pre-cached indices of a filter mask.
-    Indices(&'a [usize]),
-    // Iterator over set bits of the filter mask's boolean buffer.
-    IndicesIter(BitIndexIterator<'a>),
-    // Slice of pre-cached slices of a filter mask.
-    Slices(&'a [(usize, usize)]),
-    // Iterator over contiguous ranges of set bits of the filter mask's boolean buffer.
-    SlicesIter(arrow_buffer::bit_iterator::BitSliceIterator<'a>),
+    fn first(&self) -> Option<usize> {
+        if self.true_count == 0 {
+            return None;
+        }
+        if self.true_count == self.len {
+            return Some(0);
+        }
+        if let Some(buffer) = self.buffer.get() {
+            return buffer.set_indices().next();
+        }
+        if let Some(indices) = self.indices.get() {
+            return indices.first().copied();
+        }
+        if let Some(slices) = self.slices.get() {
+            return slices.first().map(|(start, _)| *start);
+        }
+        None
+    }
 }
 
 impl FilterMask {
     /// Create a new FilterMask where all values are set.
     pub fn new_true(length: usize) -> Self {
-        Self::from(BooleanBuffer::new_set(length))
+        Self(Arc::new(Inner {
+            buffer: Default::default(),
+            indices: Default::default(),
+            slices: Default::default(),
+            len: length,
+            true_count: length,
+            selectivity: 1.0,
+        }))
+    }
+
+    /// Create a new FilterMask where no values are set.
+    pub fn new_false(length: usize) -> Self {
+        Self(Arc::new(Inner {
+            buffer: Default::default(),
+            indices: Default::default(),
+            slices: Default::default(),
+            len: length,
+            true_count: 0,
+            selectivity: 0.0,
+        }))
+    }
+
+    /// Create a new [`FilterMask`] from a [`BooleanBuffer`].
+    pub fn from_buffer(buffer: BooleanBuffer) -> Self {
+        let true_count = buffer.count_set_bits();
+        let len = buffer.len();
+        Self(Arc::new(Inner {
+            buffer: OnceLock::from(buffer),
+            indices: Default::default(),
+            slices: Default::default(),
+            len,
+            true_count,
+            selectivity: true_count as f64 / len as f64,
+        }))
+    }
+
+    /// Create a new [`FilterMask`] from a [`Vec<usize>`].
+    pub fn from_indices(len: usize, vec: Vec<usize>) -> Self {
```

Comment:
we probably want `from_indices_unchecked` as well though we can see from profiles

Reply:
Same for slices

Reply:
I think there's something nice for these methods to still take `IntoIterator`, we could keep it as `Item = usize`

Reply:
IntoIterator makes it too easy to accidentally clone a vec I think, let's make the caller think about it.
Can add unchecked if it shows up in profiles.